### PR TITLE
Fix Existing Requests and Limits with LimitRange

### PR DESCRIPTION
**Pull request by austinzhao-go**
# Changes

This PR is to update `limitrange` logic to ensure container's requests and limits under the min and max range which is configured by limitrange.

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

``` release-note
Fixed Existing Requests and Limits with LimitRange
```

### CODE REVIEW

**Comment by austinzhao-go** [comment]
/hold 

till https://github.com/tektoncd/pipeline/pull/5082 is merged

File: pkg/internal/computeresources/transformer.go [hunk 1]
```diff
@@ -100,24 +100,38 @@ func transformPodBasedOnLimitRange(p *corev1.Pod, limitRange *corev1.LimitRange)
...
+		dst[name] = compare.MinLimit(dst[name], limits[name])
+	}
+}
+
+// setLimits sets the limits to the min of (currentLimits, defaultLimits)
+func setLimits(name corev1.ResourceName, dst corev1.ResourceList, limits corev1.ResourceList) {
+	if compare.IsZero(limits[name]) {
+		return
+	}
+	dst[name] = compare.MinLimit(dst[name], limits[name])
```
Comment (austinzhao-go) [inline pkg/internal/computeresources/transformer.go:134]: main changes related to this PR (just to exclude the depended changes in `tasklevel.go`)

File: pkg/internal/computeresources/transformer_test.go [hunk 2]
```diff
@@ -946,13 +1052,13 @@ func TestTransformerMultipleContainer(t *testing.T) {
 				Name: "sidecar-fizz",
 				Resources: corev1.ResourceRequirements{
 					Limits: corev1.ResourceList{
-						corev1.ResourceCPU:              resource.MustParse("4"),
+						corev1.ResourceCPU:              resource.MustParse("2"),
```
Comment (lbernick) [inline pkg/internal/computeresources/transformer_test.go:1055]: The LimitRange in this test case has a "default" of 2 but not a "max" of 2, so I think we want to leave the limit at 4.

File: pkg/internal/computeresources/transformer_test.go [hunk 3]
```diff
@@ -310,7 +310,7 @@ func TestTransformerOneContainer(t *testing.T) {
 			Containers: []corev1.Container{{
 				Resources: corev1.ResourceRequirements{
 					Limits: corev1.ResourceList{
-						corev1.ResourceMemory:           resource.MustParse("4Gi"),
+						corev1.ResourceMemory:           resource.MustParse("2Gi"),
```
Comment (lbernick) [inline pkg/internal/computeresources/transformer_test.go:313]: Thanks Austin! Nice job catching this bug, it looks like the behavior here wasn't consistent with our [documentation](https://github.com/tektoncd/pipeline/blob/main/docs/compute-resources.md#limits). @vdemeester could you double check me here?

File: pkg/internal/computeresources/transformer_test.go [hunk 4]
```diff
@@ -385,6 +385,112 @@ func TestTransformerOneContainer(t *testing.T) {
 				},
 			}},
 		},
+	}, {
+		description: "limitRange with min and max, while resources on containers < min",
```
Comment (lbernick) [inline pkg/internal/computeresources/transformer_test.go:389]:
```suggestion
		description: "limitRange with min and max, with requests on containers < min",
```

File: pkg/internal/computeresources/transformer_test.go [hunk 5]
```diff
@@ -385,6 +385,112 @@ func TestTransformerOneContainer(t *testing.T) {
...
+						corev1.ResourceCPU: resource.MustParse("1"),
+					},
+					Limits: corev1.ResourceList{
+						corev1.ResourceCPU: resource.MustParse("2"),
+					},
+				},
+			}},
+		},
+	}, {
+		description: "limitRange with min and max, while resources on containers > max",
```
Comment (lbernick) [inline pkg/internal/computeresources/transformer_test.go:442]:
```suggestion
		description: "limitRange with min and max, with requests on containers > max",
```

File: pkg/internal/computeresources/transformer.go [hunk 6]
```diff
@@ -100,24 +100,38 @@ func transformPodBasedOnLimitRange(p *corev1.Pod, limitRange *corev1.LimitRange)
...
 		}
 	}
 	return p
 }
 
-func setRequestsOrLimits(name corev1.ResourceName, dst, src corev1.ResourceList) {
-	if compare.IsZero(dst[name]) && !compare.IsZero(src[name]) {
-		dst[name] = src[name]
+// setRequests sets the requests to the max of (currentRequests, defaultRequests)
+// and the min of (currentRequests, defaultLimits)
```
Comment (lbernick) [inline pkg/internal/computeresources/transformer.go:118]: I think what actually needs to happen is that the requests are:
- greater than the min request
- smaller than the max limit
- set to the requests if they exist
- smaller than whatever limit we end up with

Using default limits here is not right if we have both requests and limits > default limits-- in that case we'd just want to use the specified requests and limits.

Also, this docstring reflects details of how the function is called (i.e. the "limits" are the default limits of the limitRange); could you reword it to focus on what the function does?

**Review by lbernick — COMMENTED** [review]
Thanks Austin! Would you mind splitting the limitrange commit into a separate PR?

File: pkg/internal/computeresources/transformer_test.go (hunk 3 above)
Comment (lbernick) [reply pkg/internal/computeresources/transformer_test.go:313]: I think the init container memory limits need to be changed too

File: pkg/internal/computeresources/transformer.go (hunk 6 above)
Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer.go:118]: I see. thanks for details about the logic. 

so requests will be ensured to have a value larger than Min of LimitRange while leaving limits set by the lower end of limits (of container) and LimitRange.

File: pkg/internal/computeresources/transformer_test.go (hunk 2 above)
Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:1055]: got this. so it means we actually want to overwrite the default limits whenever there are limits from containers right? 
I will update the setLimits as the "overwrite with default only if there are no limits on containers"

Comment (lbernick) [reply pkg/internal/computeresources/transformer_test.go:1055]: yes, that's right. the [docs](https://github.com/tektoncd/pipeline/blob/main/docs/compute-resources.md#limits) describe our intended behavior

Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:1055]: checked the code - the "default" in LimitRange was used to compute the `defaultContainerLimits`
https://github.com/tektoncd/pipeline/blob/99b6692a5cca86926650387830d05551236c323d/pkg/internal/computeresources/transformer.go#L193

in this case, by the doc, we still should keep the min of limits right?
so here - the defaultLimits as "4" and sidecar limits as "2", the expected is "2"

**Comment by austinzhao-go** [comment]
/hold

till https://github.com/tektoncd/pipeline/pull/5279 got merged

File: pkg/internal/computeresources/transformer_test.go (hunk 2 above)
Comment (lbernick) [reply pkg/internal/computeresources/transformer_test.go:1002]: I think it's actually the other way around here. The default limits are 2 and the sidecar limits are 4.

**Comment by lbernick** [comment]
Could you please remove the TEP title from the commit message + PR description? since this is an unrelated bug fix.

**Comment by lbernick** [comment]
Fixes #5248 
Please also add a release note!

File: pkg/internal/computeresources/transformer_test.go (hunk 2 above)
Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:1137]: your are correct. the "4" is from the sidecar. 
so the result "2" comes from the lower end of both limits.

Comment (lbernick) [reply pkg/internal/computeresources/transformer_test.go:1137]: We don't want to allow default limits to override limits set on the container. Only the limitrange max should override limits set on the container.

**Review by lbernick — COMMENTED** [review]
We also need to address the test cases "limitRange with min and limits on containers < min" and "limitRange with max and requests on containers > max". These test cases are outputting invalid pods (requests > limits).

File: pkg/internal/computeresources/transformer_test.go (hunk 2 above)
Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:1137]: right, think your understanding matches the doc we are referring. 
just the code causing me a bit confused: https://github.com/tektoncd/pipeline/blob/4c775883dcb54431574343ac85be7423711f3e72/pkg/internal/computeresources/transformer.go#L197

here the max of LimitRange is used to compute the default limit (if there is no default) - so if we only allow override limits with default, think will need to split into 2 funcs - getDefaultLimit and getMaxLimit from LimitRange - so clear the "default" here we used

Comment (lbernick) [reply pkg/internal/computeresources/transformer_test.go:1137]: got it-- let's make sure we have the test cases how we want them, and then we can change the code any way that's needed so we get the right results. If you need to split into 2 functions that's OK; for the review right now I'm just focusing on getting the test cases right.

Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:1137]: understood your logic. thanks Lee!

Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go]: kept the default as it was (since the max will be considered as the default if only max)
add a maxLimit getter func - so only take the min when max exists

**Comment by austinzhao-go** [comment]
thanks @abayer for the review!

pushed an update to ensure `requests` got "capped" by `limits` finally

File: pkg/internal/computeresources/transformer_test.go [hunk 7]
```diff
@@ -1140,7 +1140,7 @@ func TestTransformerMultipleContainer(t *testing.T) {
 					},
 					Requests: corev1.ResourceList{
 						corev1.ResourceCPU:              resource.MustParse("1"),
-						corev1.ResourceMemory:           resource.MustParse("400Mi"),
+						corev1.ResourceMemory:           resource.MustParse("200Mi"),
```
Comment (lbernick) [inline pkg/internal/computeresources/transformer_test.go:1143]: I think there was another comment thread about this, but this line should stay "400Mi".

**Comment by austinzhao-go** [comment]
/hold 

till collecting further feedback from community

**Comment by austinzhao-go** [comment]
closed the PR for this moment after a check with @lbernick 

the expected fix will be updated in another PR.
